Sources/HAP/Server/Device.swift

```diff
@@ -145,15 +200,141 @@ public class Device {
+        self.accessories = [Accessory]()
+
+        // The first accessory must be aid 1
+        accessories[0].aid = 1
+
+        addAccessories(accessories)
+    }
+
+    public func canAddAccessory(accessory: Accessory) -> Bool {
+        if !isBridge ||
+        accessories.count == 100 { // HAP Spec 2.5.3.2 - Maximum 100 accessories
+            return false
+        }
+        let serialNumber = accessory.uniqueSerialNumber
+        return isUniqueSerialNumber(serialNumber, ignoring: accessory)
+    }
+
+    // Add an array of accessories to a bridge. The result is an array of the acessories sucessfully
+    // added. An accessory will not be added if its serial number is not unique.
+    @discardableResult
+    public func addAccessories(_ newAccessories: [Accessory]) -> [Accessory] {
+        precondition(isBridge && (accessories.count + newAccessories.count) <= 100,
+                     "A maximum of 99 accessories can be added to a bridge")
+        // Remove any acessories with duplicate serial numbers
+        var verifiedAccessories = [Accessory]()
+        for accessory in newAccessories {
+            let serialNumber = accessory.uniqueSerialNumber
+            if isUniqueSerialNumber(serialNumber, ignoring: accessory) {
+                verifiedAccessories.append(accessory)
+            } else {
+                // swiftlint:disable:next line_length
+                logger.info("Accessories must have unique serial numbers. Duplicate found '\(serialNumber)'. Second device ignored")
+            }
+        }
+        if verifiedAccessories.isEmpty {
+            return []
+        }
+        accessories.append(contentsOf: verifiedAccessories)
+
+        // Check to see if the aid has been stored in the configuration data
+        for accessory in newAccessories {
             accessory.device = self
-            accessory.aid = idGenerator.next()!
+            if accessory.aid == 0 {
+                let serialNumber = accessory.uniqueSerialNumber
+                if let aid = configuration.aidForAccessorySerialNumber[serialNumber] {
+                    accessory.aid = aid
+                }
+            }
         }
+
+        // Generate new aid if one is not already found or provided
+        for accessory in newAccessories {
+            // Verify that the aid is indeed unique
+            if accessory.aid != 0,
+               !isUniqueAID(accessory.aid, ignoring: accessory) {
+                // swiftlint:disable:next line_length
+                logger.info("Accessory \(accessory.info.name.value ?? "unknown") has a duplicate accessory ID \(accessory.aid), replacing with fresh IID")
+                accessory.aid = 0
+            }
+            if accessory.aid == 0 {
+                // Obtain a new aid, which has not already been used
+                repeat {
+                    accessory.aid = configuration.nextAID()
+                } while (!isUniqueAID(accessory.aid, ignoring: accessory))
```

Comment:
Why do we need to do this? As I understand it, the configuration's last assigned AID is restored on startup. This AID will only increase, so each AID that was handed out, should then never be reused. The spec also says that the AID should be left unchanged for the remainder of the pairing. So my thinking is that we either;

* restore the AID from the configuration, if possible
* assign a new AID from the generator -- no need to check for duplicates, I think?

Reply:
The spec requires AID's to be unique. In order to avoid a HomeKit database issue, I felt it wise to aggressively ensure that. Indeed the logic is designed that an AID should be unique, however as it is relying on values stored in a configuration, which could have been corrupted or modified. Therefore I felt it was safer to ensure the uniqueness for each accessory added, even one which has been seen before.

Reply:
Ok